frontend/catalyst/api_extensions/quantum_operators.py

```diff
@@ -312,49 +315,41 @@ def cond_fn():
 
 
 ## IMPL ##
-class MidCircuitMeasure(HybridOp):
+class MidCircuitMeasure(MidMeasureMP):
     """Operation representing a mid-circuit measurement."""
 
     binder = qmeasure_p.bind
 
-    # pylint: disable=too-many-arguments
-    def __init__(
-        self,
-        in_classical_tracers,
-        out_classical_tracers,
-        regions: List[HybridOpRegion],
-        reset: bool = None,
-        postselect: int = None,
-    ):
-        HybridOp.__init__(self, in_classical_tracers, out_classical_tracers, regions)
-        self._wires = qml.wires.Wires(in_classical_tracers)
-        self.reset = reset
-        self.postselect = postselect
+    @debug_logger_init
+    def __init__(self, wires, mv, reset: bool = None, postselect: int = None):
+        super().__init__(wires=wires, reset=reset, postselect=postselect)
+        self.mv = mv
 
-    # pylint: disable=too-many-arguments
-    def trace_quantum(self, ctx, device, trace, qrp, postselect_mode=None) -> QRegPromise:
+    @debug_logger
+    def trace_quantum(self, ctx, trace, qrp, postselect_mode=None) -> QRegPromise:
+        """Perform the second, quantum part of the Hybrid operation tracing."""
         qubit = qrp.extract(self.wires)[0]
-        if postselect_mode == "hw-like":
-            qubit2 = self.bind_overwrite_classical_tracers(
-                ctx,
-                trace,
-                in_expanded_tracers=[qubit],
-                out_expanded_tracers=self.out_classical_tracers,
-            )
-        else:
-            qubit2 = self.bind_overwrite_classical_tracers(
-                ctx,
-                trace,
-                in_expanded_tracers=[qubit],
-                out_expanded_tracers=self.out_classical_tracers,
-                postselect=self.postselect,
-            )
+        kwargs = {} if postselect_mode == "hw-like" else {"postselect": self.postselect}
+        qubit2 = self.binder(qubit, **kwargs)[-1]
+
+        eqn = ctx.frames[trace].eqns[-1]
+        assert len(eqn.outvars[:-1]) == 1, f"{eqn.outvars=}\n{self.mv=}"
+
+        if trace.getvar(self.mv) not in set(
+            [
+                *sum([e.outvars for e in ctx.frames[trace].eqns[:-1]], []),
+                *ctx.frames[trace].invars,
+                *ctx.frames[trace].constvar_to_val.keys(),
+            ]
+        ):
+            eqn.outvars[0] = trace.getvar(self.mv)
```

Comment:
To me it looks like it's making `mv` the output of the current line in the jaxpr if it's not already present anywhere else in the trace. Admittedly I don't fully understand the logic or the jargon, but I get errors without this block.

Reply:
@grwlf Can you add a comment about what this block does?

Reply:
Hmm, looks like it is an approximate copy of the code residing in the `bind_overwrite_classical_tracers` [here](https://github.com/PennyLaneAI/catalyst/blob/aa19452e2c5bcf8f6aa0ee4a226daf47ff28f7b7/frontend/catalyst/jax_tracer.py#L464). This code handles a rather weird situation happening when we return dynamic dimensions from loops, so let me explain that.

Consider the following "identity-loop" program:

``` python
@qjit
@qnode(...)
def circuit(dim:int):
   @for_loop(0,10,1)
   def loop(i, a):
      return a
   b = loop(jnp.zeros([dim], float))
   ...
```

In order to trace this loop, Catalyst would parse `a` arguments into `in_type`, create inner-scope tracers, get the results, deduce `out_type` containing a reference to the input dimension and finally create an outer scope tracers for `b`. Following the InDBIdx/OutDBIdx semantics, it will re-use the `dim` tracer for the shape of `b`, which means that the expanded list of classical loop results would look something like `dim:i64[] b:f64[dim]`. If we directly interpret these results we would get a double-assignment for `dim`, e.g.

```
let dim:i64[] in # <--- First assignment
a:f64[dim] = jnp.zeros([dim], float)
dim:i64[]  b:f64[dim] =  loop({let dim:i64[] i:i64[] a:f64[dim]  in  (dim, a) }) dim a
^^^
Second assignment
```

The mentioned code avoids this by skipping the already-defined variables. We can always do this because we know that in the original program we used the same tracer in all cases.

Reply:
Speaking of this MCM code snippet, I would say that the `if var not in set(...)` condition should always fire, so my suggestion is to replace it with just
`eqn.outvars[0] = trace.getvar(self.mv)`
or, better, try calling `bind_overwrite_classical_tracers( ..., out_expanded_tracers = [self.mv], ... )`

Reply:
> Speaking of this MCM code snippet, I would say that the if var not in set(...) condition should always fire, so my suggestion is to replace it with just eqn.outvars[0] = trace.getvar(self.mv)

Nice, even better 👍 

> try calling bind_overwrite_classical_tracers( ..., out_expanded_tracers = [self.mv], ... )
>
> We might need to move it out of this HybridOp class which actually was my initial proposal.

Yes we didn't want the MCM to be a HybridOp anymore, and in case we've been wanting to move the HybridOp class out of `jax_tracer` so making the `bind_overwrite_classical_tracers` standalone might be a good idea. We'll take a look next quarter :)

Reply:
Another suggestion: can we actually use both HybridOp and MidMeasureMP as base classes for MidCircuitMeasure, rather than changing from one to another?

Reply:
> Another suggestion: can we actually use both HybridOp and MidMeasureMP as base classes for MidCircuitMeasure, rather than changing from one to another?

My preference was not to because it doesn't have a body, and reserve HybridOp for operations with nested scopes.